core/lib/vm_utils/src/storage.rs

```diff
@@ -83,18 +83,41 @@ pub fn l1_batch_params(
 
 /// Provider of L1 batch parameters for state keeper I/O implementations. The provider is stateless; i.e., it doesn't
 /// enforce a particular order of method calls.
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct L1BatchParamsProvider {
     snapshot: Option<SnapshotRecoveryStatus>,
 }
 
 impl L1BatchParamsProvider {
-    pub async fn new(storage: &mut Connection<'_, Core>) -> anyhow::Result<Self> {
-        let snapshot = storage
+    pub fn new() -> Self {
+        Self { snapshot: None }
+    }
+
+    /// Performs the provider initialization. Must only be called with the initialized storage (e.g.
+    /// either after genesis or snapshot recovery).
+    pub async fn initialize(&mut self, storage: &mut Connection<'_, Core>) -> anyhow::Result<()> {
```

Comment:
Bikeshedding: I'd rather retain the `new()` method as-is and create a new `uninitialized()` constructor (or `Default` implementation), so that it's easier to initialize the provider correctly.

Reply:
I just thought that it matches the `StateKeeperIO` well, which also has separate methods to create and initialize the object.
Actually, the fact that this structure didn't have an `initialize` method was one of the reasons why I was _extremely_ confused by the `ExternalIO` not picking up the previous batch hash and it took a while to understand what was going on.
Having a method that may either work or not depending on circumstances (which are also pretty rare) sounds a bit dangerous.